[Cocoa] Add plumbing for a couple more platform flags on ResourceRequest

whsieh opened the pull request:
#### 2092e7169114e46d8dc8ac71dad797f3c33bdc9e
<pre>
[Cocoa] Add plumbing for a couple more platform flags on ResourceRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=255927">https://bugs.webkit.org/show_bug.cgi?id=255927</a>
rdar://105895713

Reviewed by Tim Horton and Matthew Finkel.

Add plumbing for two flags on `ResourceRequest`: `PrivacyProxyFailClosedForUnreachableNonMainHosts`
and `UseNetworkConnectionIntegrity`, which both map to SPI properties on `NSURLRequest`. In a
subsequent patch, this will make it possible for WebKit clients to specify these flags on a URL
request when loading a web view, and have this state propagate to everywhere (including the resource
request used for performing preconnect).

* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setAsIsolatedCopy):
(WebCore::ResourceRequestBase::setIsAppInitiated):
(WebCore::ResourceRequestBase::setPrivacyProxyFailClosedForUnreachableNonMainHosts):
(WebCore::ResourceRequestBase::setUseNetworkConnectionIntegrity):

Add support for the new flags here; this closely follows the existing pattern used to plumb whether
or not the request is app-initiated through `ResourceRequest`.

* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::privacyProxyFailClosedForUnreachableNonMainHosts const):
(WebCore::ResourceRequestBase::useNetworkConnectionIntegrity const):
* Source/WebCore/platform/network/cf/ResourceRequest.h:
* Source/WebCore/platform/network/cf/ResourceRequestCFNet.cpp:
(WebCore::ResourceRequest::updateFromDelegatePreservingOldProperties):
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):
(WebCore::ResourceRequest::getResourceRequestPlatformData const):
(WebCore::configureRequestWithData):

Add a new helper method to set some policy flags on the platform URL request, to avoid duplicating
this logic in the two methods below.

(WebCore::ResourceRequest::doUpdatePlatformRequest):
(WebCore::ResourceRequest::doUpdatePlatformHTTPBody):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Add IPC encoding/decoding support for the platform flags.

Canonical link: <a href="https://commits.webkit.org/263390@main">https://commits.webkit.org/263390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1837a3081f1e57ddfa66616040f6f743744289a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4423 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4543 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4675 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5902 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4622 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4412 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4507 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4853 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4625 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3980 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5903 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2120 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3959 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/8032 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3960 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4028 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5567 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4430 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3607 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3954 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3958 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1095 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8005 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4309 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
```diff
@@ -339,6 +365,14 @@ static inline NSURLRequestCachePolicy toPlatformRequestCachePolicy(ResourceReque
 #endif
 
+#if HAVE(PRIVACY_PROXY_FAIL_CLOSED_FOR_UNREACHABLE_HOSTS)
+    [nsRequest _setPrivacyProxyFailClosedForUnreachableNonMainHosts:m_requestData.m_privacyProxyFailClosedForUnreachableNonMainHosts];
```
Not a problem with this patch, but we shouldn't duplicate this logic in both `doUpdatePlatformRequest` and `doUpdatePlatformHTTPBody`.

whsieh replied:
That's a great point! I'll refactor this out to a helper method. (At least, for the app attribution flag as well as these two new ones).

sysrqb reviewed (approved):
Otherwise, lgtm.

whsieh committed:
[Cocoa] Add plumbing for a couple more platform flags on ResourceRequest
https://bugs.webkit.org/show_bug.cgi?id=255927
rdar://105895713

Reviewed by Tim Horton and Matthew Finkel.

Add plumbing for two flags on `ResourceRequest`: `PrivacyProxyFailClosedForUnreachableNonMainHosts`
and `UseNetworkConnectionIntegrity`, which both map to SPI properties on `NSURLRequest`. In a
subsequent patch, this will make it possible for WebKit clients to specify these flags on a URL
request when loading a web view, and have this state propagate to everywhere (including the resource
request used for performing preconnect).

* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setAsIsolatedCopy):
(WebCore::ResourceRequestBase::setIsAppInitiated):
(WebCore::ResourceRequestBase::setPrivacyProxyFailClosedForUnreachableNonMainHosts):
(WebCore::ResourceRequestBase::setUseNetworkConnectionIntegrity):

Add support for the new flags here; this closely follows the existing pattern used to plumb whether
or not the request is app-initiated through `ResourceRequest`.

* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::privacyProxyFailClosedForUnreachableNonMainHosts const):
(WebCore::ResourceRequestBase::useNetworkConnectionIntegrity const):
* Source/WebCore/platform/network/cf/ResourceRequest.h:
* Source/WebCore/platform/network/cf/ResourceRequestCFNet.cpp:
(WebCore::ResourceRequest::updateFromDelegatePreservingOldProperties):
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):
(WebCore::ResourceRequest::getResourceRequestPlatformData const):
(WebCore::configureRequestWithData):

Add a new helper method to set some policy flags on the platform URL request, to avoid duplicating
this logic in the two methods below.

(WebCore::ResourceRequest::doUpdatePlatformRequest):
(WebCore::ResourceRequest::doUpdatePlatformHTTPBody):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Add IPC encoding/decoding support for the platform flags.

Canonical link: https://commits.webkit.org/263390@main
